src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -606,24 +606,21 @@ internal async Task<TResult> HandleInboundActivityAsync<TResult>(Func<Task<TResu
 
     private static Func<Func<Task>, Task> BuildInboundActivityDispatcher(IReadOnlyList<CircuitHandler> circuitHandlers, Circuit circuit)
     {
-        Func<CircuitInboundActivityContext, Task>? result = null;
-
-        for (var i = circuitHandlers.Count - 1; i >= 0; i--)
+        if (circuitHandlers.Count == 0)
         {
-            if (circuitHandlers[i] is IHandleCircuitActivity inboundActivityHandler)
-            {
-                var next = result ?? (static (context) => context.Handler());
-                result = (context) => inboundActivityHandler.HandleInboundActivityAsync(context, next);
-            }
+            // If there are no registered handlers, there is no need to allocate a context on each call.
+            return static handler => handler();
         }
 
-        if (result is null)
+        var result = static (CircuitInboundActivityContext context) => context.Handler();
+
+        for (var i = circuitHandlers.Count - 1; i >= 0; i--)
         {
-            // If there are no registered handlers, there is no need to allocate a context on each call.
-            return static (handler) => handler();
+            var next = result;
+            result = circuitHandlers[i].CreateInboundActivityHandler(next);
         }
 
-        return (handler) => result(new(handler, circuit));
+        return handler => result(new(handler, circuit));
     }
```

Comment:
I just remembered what you said about wanting to avoid the `CircuitInboundActivityContext` allocation when no one overrides `CreateInboundActivityHandler` in API review. To do that, you can compare the final and inner `Func<CircuitInboundActivityContext, Task>`.`


```csharp
    private static Func<Func<Task>, Task> BuildInboundActivityDispatcher(IReadOnlyList<CircuitHandler> circuitHandlers, Circuit circuit)
    {
        var inner = static (CircuitInboundActivityContext context) => context.Handler();
        var next = inner;

        for (var i = circuitHandlers.Count - 1; i >= 0; i--)
        {
            next = circuitHandlers[i].CreateInboundActivityHandler(next);
        }

        if (next == inner)
        {
            // If no handlers wrapped inner, there is no need to allocate a context on each call.
            return static handler => handler();
        }

        return handler => next(new(handler, circuit));
    }
```